runtime/shared_common/shrinit.cpp

```diff
@@ -4923,4 +4929,147 @@ isFreeDiskSpaceLow(J9JavaVM *vm, U_64* maxsize)
+ * @param[in] heapSize1  The first heap size param that is to be stored into the shared cache
+ * @param[in] heapSize2  The second heap size param that is to be stored into the shared cache
+ * @param[in] forceReplace TRUE Replace the existing GC hints under the same key (the same command line arguments) if there is one already in the shared cache.
+ * 							FALSE Do not replace existing GC hints under the same key.
+ *
+ * @return  The location of the cached GC hints or null
+ */
+const U_8*
+j9shr_storeGCHints(J9VMThread* currentThread, UDATA hintsFlags, UDATA heapSize1, UDATA heapSize2, BOOLEAN forceReplace)
+{
+	J9JavaVM* vm = currentThread->javaVM;
+	const U_8* returnVal = NULL;
+	char *key = generateGCHintsKey(vm);
+	PORT_ACCESS_FROM_JAVAVM(vm);
+	Trc_SHR_INIT_j9shr_storeGCHints_Entry(currentThread, hintsFlags, heapSize1, heapSize2, forceReplace);
+
+	if (NULL != key) {
+		J9SharedGCHintsDataDescriptor hints = {0};
+		hints.flags = hintsFlags;
+		hints.heapSize1 = heapSize1;
+		hints.heapSize2 = heapSize2;
+		J9SharedDataDescriptor dataDescriptor = {0};
+		dataDescriptor.address = (U_8*)&hints;
+		dataDescriptor.length = sizeof(J9SharedGCHintsDataDescriptor);
+		dataDescriptor.type = J9SHR_DATA_TYPE_GC_HEAPHINT_2;
+		if (TRUE == forceReplace) {
+			dataDescriptor.flags = J9SHRDATA_SINGLE_STORE_FOR_KEY_TYPE_OVERWRITE;
+		} else {
+			dataDescriptor.flags = J9SHRDATA_SINGLE_STORE_FOR_KEY_TYPE;
+		}
+		returnVal = j9shr_storeSharedData(currentThread, key, strlen(key), &dataDescriptor);
+		j9mem_free_memory(key);
+	} else {
+		Trc_SHR_INIT_j9shr_storeGCHints_Null_Key(currentThread);
+	}
+	Trc_SHR_INIT_j9shr_storeGCHints_Exit(currentThread, returnVal);
+
+	return returnVal;
+}
+
+
+/**
+ * Find the GC hints from the shared classes cache
+ * @param[in] vmThread  The current thread
+ * @param[out] hintsFlags Extra flags associated with the GC hints
+ * @param[out] heapSize1  The first heap size that has been previouly stored into the shared cache
+ * @param[out] heapSize2  The second heap size that has been previouly stored into the shared cache
+ *
+ * @return 0 on success, -1 otherwise.
+ */
+IDATA
+j9shr_findGCHints(J9VMThread* currentThread, UDATA* hintsFlags, UDATA *heapSize1, UDATA *heapSize2)
+{
+	J9JavaVM* vm = currentThread->javaVM;
+	IDATA returnVal = -1;
+	char *key = generateGCHintsKey(vm);;
+	PORT_ACCESS_FROM_JAVAVM(vm);
+	Trc_SHR_INIT_j9shr_findGCHints(currentThread);
+
+	if (NULL != key) {
+		J9SharedDataDescriptor dataDescriptor = {0};;
+		if (0 < j9shr_findSharedData(currentThread, key, strlen(key), J9SHR_DATA_TYPE_GC_HEAPHINT_2, 0, &dataDescriptor, NULL)) {
+			J9SharedGCHintsDataDescriptor hints;
+			memcpy(&hints, dataDescriptor.address, sizeof(J9SharedGCHintsDataDescriptor));
+			if (NULL != hintsFlags) {
+				*hintsFlags = hints.flags;
+			}
+			if (NULL != heapSize1) {
+				*heapSize1 = hints.heapSize1;
+			}
+			if (NULL != heapSize2) {
+				*heapSize2 = hints.heapSize2;
+			}
+			returnVal = 0;
```

Comment:
Please add a tracepoint to show the hint data found (or add it to the exit tracepoint).